### PR TITLE
Fixed onSort bug in data_table.dart

### PR DESCRIPTION
**Pull request by andermoran**
## Problem
A bad comparison argument on a ternary operator made it impossible for a null value to be passed into `_buildHeadingCell`'s `onSort` parameter causing alignment problems described in the related issues referenced below.

### Short explanation of my fix
So for the line
```dart
onSort: () => column.onSort != null ? column.onSort(dataColumnIndex, sortColumnIndex != dataColumnIndex || !sortAscending) : null
```
the comparison argument is `column.onSort != null` instead of `() => column.onSort != null`. That is where the problem lies. Because the `onSort` parameter will either take `() => column.onSort(dataColumnIndex, sortColumnIndex != dataColumnIndex || !sortAscending)` or `() => null` and **BOTH** are non-null VoidCallback functions.
If you remove the `() =>` from in front of the comparison argument, you fix that problem; however, you still need to return a callback function or null so change the second argument to `() => column.onSort(dataColumnIndex, sortColumnIndex != dataColumnIndex || !sortAscending)` so that is passes a VoidCallback function.

### Long explanation of my fix

```dart
typedef VoidCallback = void Function();

void myFunction({VoidCallback myCallback}) {
  if (myCallback != null) {
    print("myCallback is NOT null. Calling myCallback now.");
    myCallback();
  } else {
    print("myCallback is null!");
  }
}

void main() {
  VoidCallback callback0;
  VoidCallback callback1 = () => print("callback1");

  /* This is how it should be
   * -------------------------------------------
   * We check if the callback0 has been defined.
   * If it has, then we pass
   * 
   * callback0
   * 
   * as the parameter, if not, we pass
   * 
   * null
   * 
   * as the parameter.
   */
  myFunction(myCallback: callback0 != null ? callback0 : null);

  /* This is the logic data_table.dart implements
   * -------------------------------------------
   * We check if the callback0 has been defined.
   * If it has, then we pass
   * 
   * () => callback0
   * 
   * as the parameter, if not, we pass
   * 
   * () => null
   * 
   * as the parameter.
   * 
   * Regardless of whether callback0 is null or not,
   * this results in a non-null VoidCallback function
   * being passed as the parameter because both
   * 
   * () => callback0
   * 
   * and 
   * 
   * () => null
   * 
   * are non-null VoidCallback functions because
   * they are defined with that line.
   * 
   * This results in the if statement
   * 
   * if (onSort != null) [this is currently on line 435 of data_table.dart]
   * 
   * always evaluating to true.
   */
  myFunction(myCallback: () => callback0 != null ? callback0 : null);
}

```

## Related Issues
- These issues will be fixed by this pull request
  - https://github.com/flutter/flutter/issues/26845
  - https://github.com/flutter/flutter/issues/40526

## Tests

I didn't add any tests as this is already a part of Flutter's code and is an incredibly simple fix.

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require Flutter developers to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (Please read [Handling breaking changes]). *Replace this with a link to the e-mail where you asked for input on this proposed change.*
- [x] No, this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

**Comment by andermoran** [comment]
@Hixie do I need any tests for a change so straightforward?

**Comment by andermoran** [comment]
@googlebot I signed it!

**Comment by HansMuller** [comment]
This appears to be a nice fix for what was probably a typo; thanks! It will definitely need a test: we don't want this issue to crop up again.

Also: there are some presubmit DataTable test failures.

**Comment by andermoran** [comment]
@HansMuller I've never done a pull request on a public repository so I'm unsure what the next steps are to get this approved.

**Comment by willlarche** [comment]
You’re doing a great job!

 Find a test file that’s already doing tests on these classes and add one specifically to check for your case that is being fixed. 

1.

**Comment by andermoran** [comment]
So I found the test file at `flutter/packages/flutter/test/material/data_table_test.dart` and found this test

```dart
  testWidgets('DataTable column onSort test', (WidgetTester tester) async {
    await tester.pumpWidget(
      MaterialApp(
        home: Material(
          child: DataTable(
            columns: const <DataColumn>[
              DataColumn(
                label: Text('Dessert'),
              ),
            ],
            rows: const <DataRow>[
              DataRow(
                cells: <DataCell>[
                  DataCell(
                    Text('Lollipop'), // wraps
                  ),
                ],
              ),
            ],
          ),
        ),
      ),
    );
    await tester.tap(find.text('Dessert'));
    await tester.pump();
    expect(tester.takeException(), isNull);
  });
```

and I want to add a test but I'm not sure how to run this test file to ensure that whatever test I add actually works. I've only ever run dart in a flutter app. How can I run this file? Sorry if this is a dumb question but I've tried to find out how to run this test and was unable to.

**Comment by andermoran** [comment]
I know the test file is `flutter/packages/flutter/test/material/data_table_test.dart` and the test should be in the existing test

```dart
  testWidgets('DataTable column onSort test', (WidgetTester tester) async {
    await tester.pumpWidget(
      MaterialApp(
        home: Material(
          child: DataTable(
            columns: const <DataColumn>[
              DataColumn(
                label: Text('Dessert'),
              ),
            ],
            rows: const <DataRow>[
              DataRow(
                cells: <DataCell>[
                  DataCell(
                    Text('Lollipop'), // wraps
                  ),
                ],
              ),
            ],
          ),
        ),
      ),
    );
    await tester.tap(find.text('Dessert'));
    await tester.pump();
    expect(tester.takeException(), isNull);
  });
```

but I'm not sure how to write the test. I pretty much want to make sure that if no value is provided for `onSort`, then the value passed for `_buildHeadingCell`'s `onSort` parameter (on line 621 on `data_table.dart`) is null (rather than a non-null VoidCallback function, this is the current problem). But looking at the test code above that is in place for the onSort test, I am lost as to how to write a test for this. Any guidance would be very helpful.

**Comment by andermoran** [comment]
Also, how is this test

```dart
  testWidgets('DataTable column onSort test', (WidgetTester tester) async {
    await tester.pumpWidget(
      MaterialApp(
        home: Material(
          child: DataTable(
            columns: const <DataColumn>[
              DataColumn(
                label: Text('Dessert'),
              ),
            ],
            rows: const <DataRow>[
              DataRow(
                cells: <DataCell>[
                  DataCell(
                    Text('Lollipop'), // wraps
                  ),
                ],
              ),
            ],
          ),
        ),
      ),
    );
    await tester.tap(find.text('Dessert'));
    await tester.pump();
    expect(tester.takeException(), isNull);
  });
```

testing onSort? There isn't even an onSort parameter passed so what exactly is this testing for?

**Comment by andermoran** [comment]
This would be a test for DataColumn. Am I supposed to make a new data_column_test.dart?

**Comment by andermoran** [comment]
Note: the bad ternary comparison argument is also present in this snipped of code in data_table.dart but I am unaware of its side effects.

```dart
      for (DataRow row in rows) {
        tableRows[rowIndex].children[0] = _buildCheckbox(
          color: theme.accentColor,
          checked: row.selected,
          onRowTap: () => row.onSelectChanged != null ? row.onSelectChanged(!row.selected) : null ,
          onCheckboxChanged: row.onSelectChanged,
        );
        rowIndex += 1;
      }
```

**Comment by andermoran** [comment]
@HansMuller I'm stuck here and want to get this fix implemented as soon as possible! I'm struggling on the new tests I'm supposed to create (even the existing tests don't make sense) and I need some help.

**Comment by andermoran** [comment]
@HansMuller I believe this pr is ready

**Comment by HansMuller** [comment]
@andermoran the PR still doesn't include a test. I'm afraid I don't have the time to help craft one at the moment, sorry about that.

**Comment by andermoran** [comment]
Despite weeks of looking at several resources to figure out how to write flutter tests, I still can't get the hang of it. Will this bug fix ever be merged?

**Comment by Hixie** [comment]
Might https://flutter.dev/docs/cookbook/testing/widget/introduction help?

**Comment by andermoran** [comment]
@Hixie I’m not sure if you’re trolling me or not but I just said I’ve spent weeks looking at several resources to write flutter tests but have been unable to. Yes, I have looked at the intro page as well as many other resources. I’ve really tried my hardest to write these tests because I want this merge to happen badly but I’ve been unable to. I’ve looked at all the resources on the web about flutter testing as well as looking through flutter tests in flutter’s source code to how it was done there. At this point I’ve given up. I wish someone could at least write one test for this PR that I need so that I can work off that.

**Comment by Hixie** [comment]
If you could file a bug that describes what you tried and how the documentation failed you, that would be great. I'm sad to hear that the current documentation was inadequate to the task.

**Comment by deadsoul44** [comment]
Any update?

**Comment by dkwingsmt** [comment]
@andermoran I looked into it since this is indeed an interesting PR and exposes some problems of our code coverage. I came up with a unit test as below:

```dart

  testWidgets('DataTable column label position - no sort', (WidgetTester tester) async {
    Widget buildTable({ int sortColumnIndex, bool sortAscending = true }) {
      return DataTable(
        sortColumnIndex: sortColumnIndex,
        sortAscending: sortAscending,
        columns: <DataColumn>[
          const DataColumn(
            label: Expanded(child: Center(child: Text('Name'))),
            tooltip: 'Name',
          ),
          DataColumn(
            label: const Expanded(child: Center(child: Text('Calories'))),
            tooltip: 'Calories',
            numeric: true,
            onSort: (int columnIndex, bool ascending) {},
          ),
        ],
        rows: const <DataRow>[
          DataRow(
            cells: <DataCell>[
              DataCell(Text('A long desert name')),
              DataCell(Text('A lot of calories')),
            ],
          ),
        ]
      );
    }

    await tester.pumpWidget(MaterialApp(
      home: Material(child: buildTable()),
    ));

    final Finder nameText = find.text('Name');
    expect(nameText, findsOneWidget);
    final Finder nameRow = find.ancestor(of: find.text('Name'), matching: find.byType(Row)).first;
    expect(tester.getCenter(nameText), equals(tester.getCenter(nameRow)));
  });
```

This test finds the text and its immediately wrapping row, and see if they have the same center, indicating that no other padding is added. This test breaks on master and passes on your PR.

However, I did find a bug (unrelated to this PR) that causes the header to disappear on mouse move, which I've filed at https://github.com/flutter/flutter/issues/51152. This bug seems to occur only when the text is not wrapped in the `InkWell` (which is only found out thanks to your PR), so until that issue is fixed, you'll need to wrap the header in `InkWell` unconditionally.

Hopefully this is a good starting point for you!

**Comment by deadsoul44** [comment]
I fear @andermoran gave up due its frustration. One of team members from flutter should finalize the pull request. It has been already more than 5 months.

**Comment by andermoran** [comment]
> I fear @andermoran gave up due its frustration. One of team members from flutter should finalize the pull request. It has been already more than 5 months.

Can confirm. Was frustrating to create a test. Did give up.

File: packages/flutter/lib/src/material/data_table.dart [hunk 1]
```diff
@@ -443,6 +443,11 @@ class DataTable extends StatelessWidget {
         textDirection: numeric ? TextDirection.rtl : null,
         children: <Widget>[ label, arrowPadding, arrow ],
       );
+    } else {
+      label = Row(
+        textDirection: numeric ? TextDirection.rtl : null,
+        children: <Widget>[ label ],
```
Comment (Hixie) [inline packages/flutter/lib/src/material/data_table.dart]: having a row with just one child seems unnecessarily expensive... is this just to make sure that you can toggle onSort without losing state in the child? I guess that makes sense...

Probably better in that case to do this as one expression, with UI-as-code.

**Comment by Hixie** [comment]
@dkwingsmt If the original bug was alignment issues we should probably have the test toggle the sortability (have onSort be null once, then non-null) and verify that the alignment is correct in both cases.

**Comment by dkwingsmt** [comment]
@Hixie I can test that the alignment with `onSort` is consistent, but it's never gonna be at the center. (Although I can test that the row is centered in the cell, it's not useful since the row is supposed to fill up the cell).

**Comment by deadsoul44** [comment]
I would like to have datacolumn label to be still centered even if it is sorted (up or down arrow appears on the right of the centered label). It can be annoying to see centered label moves when it is sorted.

File: packages/flutter/lib/src/material/data_table.dart (hunk 1 above)
Comment (andermoran) [reply packages/flutter/lib/src/material/data_table.dart]: I see your point. I debated whether to create a whole row for just one element. I went prioritized consistency over optimization. I would have no problem if that piece of code was changed, your point is valid

**Comment by dkwingsmt** [comment]
@deadsoul44 It's a good proposal, but let me explain the obstacle. The current structure can roughly be presented as:
```dart
tableCell = Container(
  child: Row(
    children: [
      label,  // The `label` argument of `DataColumn`
      if (onSort != null) sortingArrow,
    ],
  ),
);
```

Imagine the cell is 100px wide, the arrow being 10px wide and the text 40 px wide. You would expect that:
1. If `label: Text('Cities')` the text is at the left edge of the cell, i.e. [0, 40]
2. If `label: Center(child: Text('Cities'))` the text is at the center of the cell, [30, 70]
  - This wouldn't directly work because `Row` prefers the minimum size, but you can fix it by wrapping it with an `Expand`.
3. If `label: Align(alignment: Alignment.centerRight), child: Text('Cities'))` the text is at the right edge of the possible space, i.e. [50, 90]. (You'll also need `Expand`).

Apparently No. 2 wouldn't center the text in the cell with the current implementation, because the text will at the center of the remaining space, which is [25, 65]. 
Alright, what if we add a mirroring padding at the left, so that since we know the arrow is 10px wide, make the text centered within [10, 90], which is [30, 70]? Well then No. 1 would place the text at [10, 50] instead of the desired [0, 40].

The root cause is that `DataTable` have no information of its child `label`, and can't decide what to align its child to. I'd like to hear if you have any suggestions!

**Comment by dkwingsmt** [comment]
I submitted a PR that continues this one in https://github.com/flutter/flutter/pull/51667

**Comment by deadsoul44** [comment]
I understand that it is hard to align with current structure. Another option can be to add a "labelAlignment" argument. Depending on developer choice, the label can be righr, center or left aligned. Currently, it is already assumed as right aligned if column is numeric and left aligned otherwise.

**Comment by deadsoul44** [comment]
Pending reviewer for two weeks...

**Comment by dkwingsmt** [comment]
This PR has been continued by https://github.com/flutter/flutter/pull/51667, which has been merged.

**Pull request closed without merging**